### PR TITLE
SSA time series samples

### PR DESCRIPTION
**Pull request by montebhoover**
Samples for SsaChangePointDetectorPrediction and SsaSpikeDetectorPrediction.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/SsaChangePointDetectorTransform.cs [hunk 1]
```diff
@@ -66,23 +75,127 @@ public static void SsaChangePointDetectorTransform()
...
+            // Output from ChangePoint predictions on new data:
+            // Data    Alert   Score   P-Value Martingale value
+            // 0       0     - 1.01    0.50    0.00
+            // 1       0     - 0.24    0.22    0.00
+            // 2       0     - 0.31    0.30    0.00
+            // 3       0       0.44    0.01    0.00
+            // 4       0       2.16    0.00    0.24
+            // 100     0      86.23    0.00    2076098.24
+            // 200     0     171.38    0.00    809668524.21
+            // 300     1     256.83    0.01    22130423541.93    <-- alert is on, note that delay is expected
```
Comment (codemzs) [inline docs/samples/Microsoft.ML.Samples/Dynamic/SsaChangePointDetectorTransform.cs]: >93    <-- alert is on, note that delay is expected [](start = 59, length = 50)

Can you please add the indicator when the mode was saved/loaded?

File: docs/samples/Microsoft.ML.Samples/Dynamic/SsaChangePointDetectorTransform.cs [hunk 2]
```diff
@@ -66,23 +75,127 @@ public static void SsaChangePointDetectorTransform()
...
+            for (int i = 0; i < TrainingSeasons; i++)
+                for (int j = 0; j < SeasonalitySize; j++)
+                    data.Add(new SsaChangePointData(j));
+
+            // Convert data to IDataView.
+            var dataView = ml.CreateStreamingDataView(data);
+
+            // Setup SsaChangePointDetector arguments
+            string outputColumnName = "Prediction";
+            string inputColumnName = "Value";
```
Comment (codemzs) [inline docs/samples/Microsoft.ML.Samples/Dynamic/SsaChangePointDetectorTransform.cs]: >"Value"; [](start = 37, length = 8)

nameof(SsaChangePointData.Value) and same for Prediction as well for both the files.

Comment (montebhoover) [reply docs/samples/Microsoft.ML.Samples/Dynamic/SsaChangePointDetectorTransform.cs]: Thanks! Done.

---
In reply to: [238001359](https://github.com/dotnet/machinelearning/pull/1788#discussion_r238001359) [](ancestors = 238001359)

File: docs/samples/Microsoft.ML.Samples/Dynamic/SsaChangePointDetectorTransform.cs (hunk 1 above)
Comment (montebhoover) [reply docs/samples/Microsoft.ML.Samples/Dynamic/SsaChangePointDetectorTransform.cs]: Good idea, I'll add that.

---
In reply to: [238001208](https://github.com/dotnet/machinelearning/pull/1788#discussion_r238001208) [](ancestors = 238001208)

File: docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs [hunk 3]
```diff
@@ -27,36 +31,41 @@ class SsaSpikePrediction
...
-            for (int i = 0; i < size / 2; i++)
-                data.Add(new SsaSpikeData(5));
+            // Generate sample series data with a recurring pattern and a spike within the pattern
+            const int SeasonalitySize = 5;
+            const int TrainingSeasons = 3;
+            const int TrainingSize = SeasonalitySize * TrainingSeasons;
+            var data = new List<SsaSpikeData>();
+            for (int i = 0; i < TrainingSeasons; i++)
+                for (int j = 0; j < SeasonalitySize; j++)
+                    data.Add(new SsaSpikeData(j));
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs]: Data generation used accross samples consider moving them in the Microsoft.ML.SamplesUtils.DatasetUtils.

File: docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs [hunk 4]
```diff
@@ -66,30 +75,135 @@ public static void SsaSpikeDetectorTransform()
...
+            // 4       0       1.33    0.18
+            // 0       0     - 1.79    0.07
+            // 1       0       0.16    0.50
+            // 2       0       0.09    0.50
+            // 3       0       0.08    0.45
+            // 4       0       1.31    0.12
+            // 100     1      98.21    0.00   <-- alert is on, predicted spike
+            // 0       0    - 13.83    0.29
+            // 1       0     - 1.74    0.44
+            // 2       0     - 0.47    0.46
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs]: doe it make sense to limit the preview/data to something smaller,maybe like 7  rows ?

File: docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs [hunk 5]
```diff
@@ -66,30 +75,135 @@ public static void SsaSpikeDetectorTransform()
             var predictionColumn = transformedData.AsEnumerable<SsaSpikePrediction>(ml, reuseRowObject: false);
 
             Console.WriteLine($"{outputColumnName} column obtained post-transformation.");
-            Console.WriteLine("Alert\tScore\tP-Value");
+            Console.WriteLine("Data\tAlert\tScore\tP-Value");
+            int k = 0;
             foreach (var prediction in predictionColumn)
-                Console.WriteLine("{0}\t{1:0.00}\t{2:0.00}", prediction.Prediction[0], prediction.Prediction[1], prediction.Prediction[2]);
+                Console.WriteLine("{0}\t{1}\t{2:0.00}\t{3:0.00}", data[k++].Value, prediction.Prediction[0], prediction.Prediction[1], prediction.Prediction[2]);
             Console.WriteLine("");
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs]: > Console.WriteLine(""); [](start = 11, length = 23)

remove, and include the newline in the above.

File: docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs [hunk 6]
```diff
@@ -66,30 +75,135 @@ public static void SsaSpikeDetectorTransform()
...
+            // 0       0    - 13.83    0.29
+            // 1       0     - 1.74    0.44
+            // 2       0     - 0.47    0.46
+            // 3       0    - 16.50    0.29
+            // 4       0    - 29.82    0.21
+        }
+
+        // This example shows spike detection as above, but demonstrates how to train a model
+        // that can run predictions on streaming data, and how to persist the trained model and then re-load it.
+        public static void SsaSpikeDetectorPrediction()
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/SsaSpikeDetectorTransform.cs]: same comment as in the other PR; maybe the end-to-end should go on tutorials. The API call usage should go here.

**Pull request merged**